File: android/app/src/main/java/com/runbeat/pro/PreferencesManager.java
package com.runbeat.pro;

import android.content.Context;
import android.content.SharedPreferences;

public class PreferencesManager {

    private static final String PREF_NAME = "runbeat_prefs";

    // KEY_BPM_LEGACY: 旧版以 float 存储的键（只读，用于迁移）
    private static final String KEY_BPM_LEGACY  = "bpm";
    // KEY_BPM_BITS: 新版以 long 存储 double 位模式，零精度损失
    private static final String KEY_BPM_BITS    = "bpm_bits";
    private static final String KEY_TICK_VOLUME = "tick_volume";
    private static final String KEY_ACCENT      = "accent";
    private static final String KEY_TIMBRE      = "timbre";
    private static final String KEY_GAIN_LEVEL  = "gain_level";
    private static final String KEY_ACCENT_INDEX = "accent_index";
    private static final String KEY_LANGUAGE     = "language";

    private static final double  DEFAULT_BPM         = 180.0;
    private static final int     DEFAULT_TICK_VOLUME  = 80;
    private static final boolean DEFAULT_ACCENT       = true;
    private static final int     DEFAULT_TIMBRE       = 0;
    private static final int     DEFAULT_GAIN_LEVEL   = 1;
    private static final int     DEFAULT_ACCENT_INDEX = 0;

    private final SharedPreferences prefs;

    public PreferencesManager(Context context) {
        prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
    }

    /**
     * 读取 BPM。优先读取新版 long bits 键；若不存在则从旧版 float 键迁移（一次性）。
     */
    public double getBpm() {
        if (prefs.contains(KEY_BPM_BITS)) {
            return Double.longBitsToDouble(
                    prefs.getLong(KEY_BPM_BITS, Double.doubleToRawLongBits(DEFAULT_BPM)));
        }
        // 旧版迁移：读取旧 float 值并立即写入新格式
        double legacy = prefs.getFloat(KEY_BPM_LEGACY, (float) DEFAULT_BPM);
        setBpm(legacy); // 写入新键，下次直接读新键
        return legacy;
    }

    /**
     * 存储 BPM（以 Double.doubleToRawLongBits 存为 long，保留完整双精度精度）。
     */
    public void setBpm(double bpm) {
        prefs.edit()
                .putLong(KEY_BPM_BITS, Double.doubleToRawLongBits(bpm))
                .apply();
    }

    public int getTickVolume() {
        return prefs.getInt(KEY_TICK_VOLUME, DEFAULT_TICK_VOLUME);
    }

    public void setTickVolume(int volume) {
        prefs.edit().putInt(KEY_TICK_VOLUME, volume).apply();
    }

    public boolean isAccentEnabled() {
        return prefs.getBoolean(KEY_ACCENT, DEFAULT_ACCENT);
    }

    public void setAccentEnabled(boolean enabled) {
        prefs.edit().putBoolean(KEY_ACCENT, enabled).apply();
    }

    public int getTimbre() {
        return prefs.getInt(KEY_TIMBRE, DEFAULT_TIMBRE);
    }

    public void setTimbre(int index) {
        prefs.edit().putInt(KEY_TIMBRE, index).apply();
    }

    public int getGainLevel() {
        int v = prefs.getInt(KEY_GAIN_LEVEL, DEFAULT_GAIN_LEVEL);
        if (v < 1) return 1;
        if (v > 3) return 3;
        return v;
    }

    public void setGainLevel(int level) {
        prefs.edit().putInt(KEY_GAIN_LEVEL, level).apply();
    }

    public int getAccentIndex() {
        int v = prefs.getInt(KEY_ACCENT_INDEX, DEFAULT_ACCENT_INDEX);
        return (v >= 0 && v < 5) ? v : DEFAULT_ACCENT_INDEX;
    }

    public void setAccentIndex(int index) {
        prefs.edit().putInt(KEY_ACCENT_INDEX, index).apply();
    }

    public String getLanguage() {
        return prefs.getString(KEY_LANGUAGE, "zh");
    }

    public void setLanguage(String lang) {
        prefs.edit().putString(KEY_LANGUAGE, lang).commit();
    }
}

File: ios/RunBeatPro/Service/PreferencesManager.swift
import Foundation

/// 偏好设置管理器 — 封装 UserDefaults（对标 Android PreferencesManager）
/// 所有持久化操作委托给 AppState 的 didSet，此处提供辅助方法
struct PreferencesManager {

    // MARK: - Keys

    private enum Key {
        static let bpmBits       = "bpm_bits"
        static let tickVolume    = "tick_volume"
        static let accent        = "accent"
        static let timbre        = "timbre"
        static let gainLevel     = "gain_level"
        static let accentIndex   = "accent_index"
        static let language      = "language"
    }

    private let ud = UserDefaults.standard

    // MARK: - BPM

    var bpm: Double {
        get {
            let v = ud.double(forKey: Key.bpmBits)
            return v > 0 ? v : 180.0
        }
        set {
            ud.set(newValue, forKey: Key.bpmBits)
        }
    }

    // MARK: - Tick Volume

    var tickVolume: Int {
        get { ud.integer(forKey: Key.tickVolume) }
        set { ud.setValue(newValue, forKey: Key.tickVolume) }
    }

    // MARK: - Accent

    var accentEnabled: Bool {
        get { ud.object(forKey: Key.accent) as? Bool ?? true }
        set { ud.setValue(newValue, forKey: Key.accent) }
    }

    // MARK: - Timbre

    var timbreIndex: Int {
        get { ud.integer(forKey: Key.timbre) }
        set { ud.setValue(newValue, forKey: Key.timbre) }
    }

    // MARK: - Gain Level

    var gainLevel: Int {
        get {
            let v = ud.integer(forKey: Key.gainLevel)
            return (v >= 1 && v <= 3) ? v : 1
        }
        set { ud.setValue(newValue, forKey: Key.gainLevel) }
    }

    // MARK: - Accent Color Index

    var accentColorIndex: Int {
        get {
            let v = ud.integer(forKey: Key.accentIndex)
            return (v >= 0 && v < 5) ? v : 0
        }
        set { ud.setValue(newValue, forKey: Key.accentIndex) }
    }

    // MARK: - Language

    var language: String {
        get { ud.string(forKey: Key.language) ?? "zh" }
        set { ud.setValue(newValue, forKey: Key.language) }
    }
}
